Row.c

```diff
@@ -468,28 +468,49 @@ void Row_printRate(RichString* str, double rate, bool coloring) {
    }
 
    if (!isNonnegative(rate)) {
-      RichString_appendAscii(str, shadowColor, "        N/A ");
-   } else if (rate < 0.005) {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f B/s ", rate);
-      RichString_appendnAscii(str, shadowColor, buffer, len);
+      RichString_appendAscii(str, shadowColor, "     N/A ");
    } else if (rate < ONE_K) {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f B/s ", rate);
-      RichString_appendnAscii(str, baseColor, buffer, len);
-   } else if (rate < ONE_M) {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f K/s ", rate / ONE_K);
-      RichString_appendnAscii(str, baseColor, buffer, len);
-   } else if (rate < ONE_G) {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f M/s ", rate / ONE_M);
-      RichString_appendnAscii(str, megabytesColor, buffer, len);
-   } else if (rate < ONE_T) {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f G/s ", rate / ONE_G);
-      RichString_appendnAscii(str, largeNumberColor, buffer, len);
-   } else if (rate < ONE_P) {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f T/s ", rate / ONE_T);
-      RichString_appendnAscii(str, largeNumberColor, buffer, len);
+      int len = snprintf(buffer, sizeof(buffer), "%4.0f B/s ", rate);
+      RichString_appendnAscii(str, shadowColor, buffer, len);
    } else {
-      int len = snprintf(buffer, sizeof(buffer), "%7.2f P/s ", rate / ONE_P);
-      RichString_appendnAscii(str, largeNumberColor, buffer, len);
+      size_t unitPrefixIndex = 0;
+      do {
+         if (unitPrefixIndex > ARRAYSIZE(unitPrefixes)-1) {
+            int len = snprintf(buffer, sizeof(buffer), "     INF ");
+            RichString_appendnAscii(str, largeNumberColor, buffer, len);
+            return;
+         }
+         unitPrefixIndex++;
+         rate /= ONE_K;
+      } while (rate >= ONE_K);
+
+      unitPrefixIndex--; // unitPrefixes starts from K
+
+      int precision = 0;
+      if (rate <= 999.9) {
+         if (rate <= 99.9) {
+            if (rate <= 9.99)
+               precision = 3;
+            else
+               precision = 2;
+         }
+         else
+            precision = 1;
+      }
+
+      if (precision < 3) {
+         double upper_limit = (precision < 2) ? ((precision < 1) ? 1000 : 100) : 10;
+         if (rate < upper_limit) {
+            rate = upper_limit;
+         }
+      }
+
+      int len = snprintf(buffer, sizeof(buffer), "%5.*f", precision, rate);
```

Comment:
I thought you would want `"%4.*f"` as the format. That is, one digit less.

Output example A
```text
   0 B/s - 9999 B/s
9.77 K/s - 9.99 K/s
10.0 K/s - 99.9 K/s
 100 K/s - 9999 K/s
9.77 M/s - 9.99 M/s
     ...
 100 Q/s - 9999 Q/s
infinity
```

Output example B
```text
   0 B/s - 1023 B/s
1.00 K/s - 9.99 K/s
10.0 K/s - 99.9 K/s
 100 K/s - 1023 K/s
1.00 M/s - 9.99 M/s
     ...
 100 Q/s - 9999 Q/s
infinity
```

Reply:
we actually want 4 digits.
By using `"%5.*f"`, it meant the total width of the output field is 5, including the decimal point and digits.

Reply:
Then you would get `999.9 K/s`. I bet you didn't test that one out.

Reply:
As its yes, should I shift to `4.*f` now, or use `5.*f`. 
I am asking this because for `4.*f` I have to change the precision logic and if it stays `5.*f` I have to add a column width, currently its 9 (for any kind of Disk IO rates), it will be then 10.

Reply:
> Then you would get `999.9 K/s`. I bet you didn't test that one out.

I wanted to understand, is there any issue with printing `999.9 K/s`? I am unable to get it...

Reply:
For size units we usually just use 3 significant digits as extra precisions are less useful for UI display purpose. Add a decimal point and it becomes 4 characters of width. When the decimal points are not needed, i.e. for numbers in range 100 - 9999, four digits can be shown for the same width, but that's more of a special case.

The numbers jump from one unit to the next in power of 1024, and that was the reason of 3 significant digits.